arbnode/message_pruner.go

```diff
@@ -25,107 +28,206 @@ type MessagePruner struct {
 	stopwaiter.StopWaiter
 	transactionStreamer *TransactionStreamer
 	inboxTracker        *InboxTracker
-	staker              *staker.Staker
 	config              MessagePrunerConfigFetcher
+	pruningLock         sync.Mutex
+	lastPruneDone       time.Time
 }
 
 type MessagePrunerConfig struct {
-	Enable               bool          `koanf:"enable"`
-	MessagePruneInterval time.Duration `koanf:"prune-interval" reload:"hot"`
+	Enable                 bool          `koanf:"enable"`
+	MessagePruneInterval   time.Duration `koanf:"prune-interval" reload:"hot"`
+	SearchBatchReportLimit int64         `koanf:"search-batch-report" reload:"hot"`
+	MinBatchesLeft         uint64        `koanf:"min-batches-left" reload:"hot"`
 }
 
 type MessagePrunerConfigFetcher func() *MessagePrunerConfig
 
 var DefaultMessagePrunerConfig = MessagePrunerConfig{
-	Enable:               true,
-	MessagePruneInterval: time.Minute,
+	Enable:                 true,
+	MessagePruneInterval:   time.Minute,
+	SearchBatchReportLimit: 100000,
+	MinBatchesLeft:         2,
 }
 
 func MessagePrunerConfigAddOptions(prefix string, f *flag.FlagSet) {
 	f.Bool(prefix+".enable", DefaultMessagePrunerConfig.Enable, "enable message pruning")
 	f.Duration(prefix+".prune-interval", DefaultMessagePrunerConfig.MessagePruneInterval, "interval for running message pruner")
+	f.Int64(prefix+".search-batch-report", DefaultMessagePrunerConfig.SearchBatchReportLimit, "limit for searching for a batch report when pruning (negative disables)")
+	f.Uint64(prefix+".min-batches-left", DefaultMessagePrunerConfig.MinBatchesLeft, "min number of batches not pruned")
 }
 
-func NewMessagePruner(transactionStreamer *TransactionStreamer, inboxTracker *InboxTracker, staker *staker.Staker, config MessagePrunerConfigFetcher) *MessagePruner {
+func NewMessagePruner(transactionStreamer *TransactionStreamer, inboxTracker *InboxTracker, config MessagePrunerConfigFetcher) *MessagePruner {
 	return &MessagePruner{
 		transactionStreamer: transactionStreamer,
 		inboxTracker:        inboxTracker,
-		staker:              staker,
 		config:              config,
 	}
 }
 
 func (m *MessagePruner) Start(ctxIn context.Context) {
 	m.StopWaiter.Start(ctxIn, m)
-	m.CallIteratively(m.prune)
 }
 
-func (m *MessagePruner) prune(ctx context.Context) time.Duration {
-	latestConfirmedNode, err := m.staker.Rollup().LatestConfirmed(
-		&bind.CallOpts{
-			Context:     ctx,
-			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
-		})
+func (m *MessagePruner) UpdateLatestStaked(count arbutil.MessageIndex, globalState validator.GoGlobalState) {
+	locked := m.pruningLock.TryLock()
+	if !locked {
+		return
+	}
+
+	if m.lastPruneDone.Add(m.config().MessagePruneInterval).After(time.Now()) {
+		m.pruningLock.Unlock()
+		return
+	}
+	err := m.LaunchThreadSafe(func(ctx context.Context) {
+		defer m.pruningLock.Unlock()
+		err := m.prune(ctx, count, globalState)
+		if err != nil {
+			log.Error("error while pruning", "err", err)
```

Comment:
We should not log this error if it's a `context.Canceled` (otherwise we'd often get that on shutdown)